src/core/distributions/events/MarkovTimesDistribution.cpp

```diff
@@ -0,0 +1,256 @@
+{
+	// simulate the new event times
+	this->setValue( simulate() );
+
+	// simulate children
+	this->simulateChildren();
+}
+
+void MarkovTimesDistribution::executeMethod(const std::string &n, const std::vector<const DagNode *> &args, long &rv) const
+{
+    if ( n == "getNumberOfEvents" )
+    {
+        rv = value->size();
+    }
+    else
+    {
+        throw RbException("The markov event model does not have a member method called '" + n + "'.");
+    }
+}
+
+void MarkovTimesDistribution::executeMethod(const std::string &n, const std::vector<const DagNode *> &args, RbVector<double> &rv) const
+{
+    if ( n == "getTimes" )
+    {
+    	// get the times as a set
+    	const std::set<double>& times = value->getEventTimes();
+
+    	// convert the times to a vector
+    	std::vector<double> res;
+    	for( std::set<double>::const_iterator it = times.begin(); it != times.end(); ++it)
+    	{
+    		res.push_back( *it );
+    	}
+
+    	// add the times to the rv
+    	rv = res;
+    }
+    else
+    {
+        throw RbException("The markov event model does not have a member method called '" + n + "'.");
+    }
+}
+
+void MarkovTimesDistribution::simulateEventTime(double &time, double &ln_prob)
+{
+	// get a random number generator
+    RandomNumberGenerator* rng = GLOBAL_RNG;
+
+    // compute the boundaries
+	const double& max = age->getValue();
+
+	// generate the new age
+	double new_age = rng->uniform01() * max;
+
+	// store the values
+	time = new_age;
+
+	// store the probability
+	ln_prob = -std::log(max);
+}
+
+void MarkovTimesDistribution::getRandomTime(double &time, double &ln_prob)
+{
+	// get a random number generator
+    RandomNumberGenerator* rng = GLOBAL_RNG;
+
+    // get the event times
+	const std::set<double>& event_times = this->value->getEventTimes();
+
+    // choose an element to update
+    size_t num_events = this->value->size();
+	size_t index = size_t(rng->uniform01() * num_events);
+
+	// get the value of the chosen element
+	std::set<double>::const_iterator the_event = event_times.begin();
+	std::advance(the_event, index);
+
+	// store the time
+	time = *the_event;
+
+    // compute the boundaries
+	const double& max = age->getValue();
+
+	// store the probability
+	ln_prob = -std::log(max);
+}
+
+void MarkovTimesDistribution::removeTime(double time)
+{
+	// remove the event
+	bool success = this->value->removeEvent(time);
+	if ( success == false )
+	{
+		throw RbException("Failed to remove time.");
+	}
+}
+
+void MarkovTimesDistribution::addTime(double time)
+{
+	// remove the event
+	bool success = this->value->addEvent(time);
+	if ( success == false )
+	{
+		throw RbException("Failed to add time.");
+	}
+}
+
+void MarkovTimesDistribution::simulateChildren()
+{
+	// check whether children need to be simulated
+	const std::vector<DagNode*>& children = this->dag_node->getChildren();
+	for(std::vector<DagNode*>::const_iterator it = children.begin(); it != children.end(); ++it)
```

Comment:
Again, this might be more readable as `for(auto& child: this->dag_node->children)`.
Then you can replace `(*it)->` with just `child.`.